dash/_hooks.py

```diff
@@ -0,0 +1,205 @@
+import typing as _t
+from importlib import metadata as _importlib_metadata
+
+import flask as _f
+
+from .exceptions import HookError
+from .resources import ResourceType
+
+if _t.TYPE_CHECKING:
+    from .dash import Dash
+    from .development.base_component import Component
+
+    ComponentType = _t.TypeVar("ComponentType", bound=Component)
+    LayoutType = _t.Union[ComponentType, _t.List[ComponentType]]
+else:
+    LayoutType = None
+    ComponentType = None
+    Dash = None
+
+
+# pylint: disable=too-few-public-methods
+class _Hook:
+    def __init__(self, func, priority, final=False, data=None):
+        self.func = func
+        self.final = final
+        self.data = data
+        self.priority = priority
+
+    def __call__(self, *args, **kwargs):
+        return self.func(*args, **kwargs)
+
+
+class _Hooks:
+    def __init__(self) -> None:
+        self._ns = {
+            "setup": [],
+            "layout": [],
+            "routes": [],
+            "error": [],
+            "callback": [],
+            "index": [],
+        }
+        self._js_dist = []
+        self._css_dist = []
+        self._finals = {}
+
+    def add_hook(
+        self, hook: str, func: _t.Callable, priority=None, final=False, data=None
+    ):
+        if final:
+            existing = self._finals.get(hook)
+            if existing:
+                raise HookError("Final hook already present")
+            self._finals[hook] = _Hook(func, priority, final, data=data)
+            return
+        hks = self._ns.get(hook, [])
+        if not priority and len(hks):
+            priority_max = max(h.priority for h in hks)
+            priority = priority_max - 1
+        elif not priority:
+            priority = 0
+        hks.append(_Hook(func, priority=priority, data=data))
+        self._ns[hook] = sorted(hks, reverse=True, key=lambda h: h.priority)
+
+    def get_hooks(self, hook: str) -> _t.List[_Hook]:
+        final = self._finals.get(hook, None)
+        if final:
+            final = [final]
+        else:
+            final = []
+        return self._ns.get(hook, []) + final
```

Comment:
Should a user of `hooks` call these two functions directly? If so, what would be the reason to do that?

Reply:
No, they should not be called by the users directly.